Add tracks to products list page

joshuatf opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Adds tracks to the products page to better understand the features that are being used.

Closes #32671  .

### How to test the changes in this Pull Request:

1.  Open your browser console and run `localStorage.setItem( 'debug', 'wc-admin:*' );`
2.  Navigate to the products page
3.  Check that the below events fire with your expected values
4.  Navigate to non-product pages and make sure the `product-tracking` script is not loaded on those pages by viewing the page's source code.

**Tracks:**
* Filter products - Select some filter dropdown options and click "Filter"
* Bulk action - Select some product checkboxes and click "Apply" after selecting a bulk action
* Bulk update - Click "Update" after selecting some fields to bulk update
* Action buttons - Hover a product row and click each of the actions that appear, (Edit, Quick Edit, Trash, etc)
* Feature button - Toggle a product as featured using the star icon
* Search - Type a search query and click "Search products"
* Sort columns - Sort the header columns both ascending and descending

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?
* [ ] Have you created a changelog file by running `pnpm nx affected --target=changelog`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

joshuatf committed:
Add tracks to products list page

joshuatf committed:
Remove duplicate tracks

joshuatf committed:
Add changelog entry

louwie17 commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-tracking/index.js:
```diff
@@ -0,0 +1,132 @@
+		sale:
+			document.querySelector( '[name="change_sale_price"]' ).value !== '',
+		tax_status:
+			document.querySelector( '[name="_tax_status"]' ).value !== '',
```
I am getting an error here, where `document.querySelector( '[name="_tax_status"]' )` returns null, so it can't find the value.
I don't see this field either, maybe it gets displayed only on specific occasions?
We probably want to change it to document.querySelector( '[name="_tax_status"]' )?.value, maybe we should do that for all the fields here?
<img width="1303" alt="Screen Shot 2022-05-16 at 1 52 23 PM" src="https://user-images.githubusercontent.com/2240960/168644869-30b638c2-cbf8-4e5f-9329-cbf230eaa9b5.png">

joshuatf replied:
Good catch on all these.  I added a function to handle checking element existence and if the value was not set (i.e., `''` or `'-1'`).

louwie17 commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-tracking/index.js:
```diff
@@ -0,0 +1,132 @@
+} );
+
+document.querySelector( '#bulk_edit' ).addEventListener( 'click', function () {
+	recordEvent( 'products_list_bulk_edit_update', {
```
It might be worth to also add a recordEvent for when the user clicks `Cancel` on the bulk edit screen.
Maybe we could use `products_list_bulk_edit_cancel`? I know this isn't part of the original issue.

joshuatf replied:
Good idea.  Added in https://github.com/woocommerce/woocommerce/pull/32949/commits/cd9c5048a673f4d6a27359e2dd23faaa4dc52523.

louwie17 commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-tracking/index.js:
```diff
@@ -0,0 +1,132 @@
+	} );
+} );
+
+bulkActionsButton.addEventListener( 'click', function () {
```
I get an error in my console here when the product list is empty (in my case I filtered by variable products of which I didn't have any).
<img width="1318" alt="Screen Shot 2022-05-16 at 1 58 13 PM" src="https://user-images.githubusercontent.com/2240960/168645145-70349d47-88b4-4eea-bd5d-a15128bfca88.png">

joshuatf replied:
Added a check for all elements before binding listeners since it's possible these could be removed/modified by plugins as well.

louwie17 commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-tracking/index.js:
```diff
@@ -0,0 +1,132 @@
+
+filterButton.addEventListener( 'click', function () {
+	recordEvent( 'products_list_filter_click', {
+		search_string_length: searchInput.value.length,
```
I get an error here after I filtered by a product type that resulted in an empty list, I then reset the product type filter and hit `Filter` again.
In an empty list the `searchInput` isn't present, so it errors on `searchInput.value.length`.

<img width="1313" alt="Screen Shot 2022-05-16 at 2 01 05 PM" src="https://user-images.githubusercontent.com/2240960/168645484-b9c2ae12-5fcf-4bf0-b00e-46dbc7b32f47.png">

joshuatf replied:
👍

louwie17 reviewed (changes requested):
Overall nice work, this tested well for the most part. I did run into a few JS errors as certain fields are hidden at times.
It might be worth being extra cautious here and heavily make use of `?.` to make sure all tracks do get triggered correctly.

I also left a suggestion of adding an extra track for cancelling out of the bulk edit screen.

Lastly, I was curious if there was a particular reason you handled all the tracks on the front-end? I figured it might be hard to differentiate between filtering, searching, and sorting columns on the back-end, meaning that the JS way would be more accurate?
Not a blocker for me, mostly just curious.

joshuatf committed:
Check for fields before getting values

joshuatf committed:
Add bulk actions cancel button

joshuatf committed:
Check for element before binding event listener

joshuatf committed:
Check for search input before getting value

joshuatf commented:
Thanks for the feedback, @louwie17!  I think I addressed all your comments.

>Lastly, I was curious if there was a particular reason you handled all the tracks on the front-end? 

After our previous Zoom discussion, I think we had to rule out using the backend since we couldn't feasibly determine which action the user most recently took (filters and search vars being set simultaneously).

Also since we also make use of filters via direct URLs, this could give us false positives on filter usage as well.

botwoo commented:
:bar_chart: Test reports for 6a22541c892d7ecf65b415ba8b19cc2a725bb0b1 `Check for search input before getting value` have been published and are accessible through the following links:

- [REST API Test Report](https://woocommerce.github.io/woocommerce-test-reports/pr/32949/api/)
- E2E Test Report (TBC)

Visit the [WooCommerce Test Reports](https://woocommerce.github.io/woocommerce-test-reports) homepage to view all published reports. See the [FAQs](https://github.com/woocommerce/woocommerce-test-reports/wiki/FAQs) page if you're having problems accessing them.

louwie17 reviewed (approved):
Thanks for the updated changes, it tested well this time around and changes look great, this is good to 🚢 

> After our previous Zoom discussion, I think we had to rule out using the backend since we couldn't feasibly determine which action the user most recently took (filters and search vars being set simultaneously).
> Also since we also make use of filters via direct URLs, this could give us false positives on filter usage as well.

Gotcha, thanks for the extra clarification, that makes sense!
